turbo-txpool/src/lib.rs

```diff
@@ -1,37 +1,178 @@
+pub mod error;
+mod tx;
+
+use crate::error::Error;
+use crate::tx::{Priced, Tx};
+
+use ethereum_types::H256;
+
+use slab::Slab;
+
+use std::collections::{hash_map, BTreeSet, HashMap};
 use std::future::Future;
 use std::pin::Pin;
 
+use tokio::sync::RwLock;
+
 use tonic::{Request, Response, Status};
 
 use turbo_proto::tonic;
 use turbo_proto::txpool::txpool_server as server;
 use turbo_proto::txpool::{
     GetTransactionsReply, GetTransactionsRequest, ImportReply, ImportRequest,
-    TxHashes,
+    ImportResult, TxHashes,
 };
 
-pub struct TxPool {}
+use typed_builder::TypedBuilder;
+
+#[derive(Debug)]
+struct Inner {
+    txs: Slab<Tx>,
+    by_hash: HashMap<H256, usize>,
+    by_price: BTreeSet<Priced>,
+
+    max_txs: usize,
+}
+
+impl Inner {
+    fn cheapest_key(&self) -> Option<usize> {
+        self.by_price.iter().next().map(|p| p.key)
+    }
+
+    fn cheapest(&self) -> Option<&Tx> {
+        self.cheapest_key().map(|k| &self.txs[k])
+    }
+
+    fn insert(&mut self, tx: Tx) -> ImportResult {
+        if self.txs.len() >= self.max_txs {
+            let cheapest =
+                self.cheapest().map(|t| t.gas_price).unwrap_or_default();
+            if tx.gas_price <= cheapest {
+                return ImportResult::FeeTooLow;
+            }
+        }
+
+        let by_hash = match self.by_hash.entry(tx.hash) {
+            hash_map::Entry::Vacant(v) => v,
+            hash_map::Entry::Occupied(_) => return ImportResult::AlreadyExists,
+        };
+
+        let gas_price = tx.gas_price;
+        let key = self.txs.insert(tx);
+
+        let inserted = self.by_price.insert(Priced { gas_price, key });
+        assert!(inserted);
+
+        by_hash.insert(key);
+
+        while self.txs.len() > self.max_txs {
+            self.remove(self.cheapest_key().unwrap());
+        }
+
+        ImportResult::Success
+    }
+
+    fn remove(&mut self, key: usize) {
+        let tx = self.txs.remove(key);
+        self.by_hash.remove(&tx.hash).expect("desync by hash");
+
+        let removed = self.by_price.remove(&Priced {
+            gas_price: tx.gas_price,
+            key,
+        });
+        assert!(removed, "desync by price");
+    }
+
+    pub fn with_config(config: Config) -> Self {
+        Self {
+            max_txs: config.max_txs,
+            txs: Slab::with_capacity(config.max_txs),
+            by_hash: HashMap::with_capacity(config.max_txs),
+            by_price: BTreeSet::new(),
+        }
+    }
+
+    pub fn find_unknown_transactions(
```

Comment:
I know this isn't our naming convention, but IMO it's rather confusing whether the txs are unknown to this client or the client's peer.